src/init.cpp

```diff
@@ -1670,12 +1670,9 @@ bool AppInitMain(InitInterfaces& interfaces)
         }
     }
 
-    if (chainparams.GetConsensus().vDeployments[Consensus::DEPLOYMENT_SEGWIT].nTimeout != 0) {
-        // Only advertise witness capabilities if they have a reasonable start time.
-        // This allows us to have the code merged without a defined softfork, by setting its
-        // end time to 0.
-        // Note that setting NODE_WITNESS is never required: the only downside from not
-        // doing so is that after activation, no upgraded nodes will fetch from you.
+    if (chainparams.GetConsensus().SegwitHeight != std::numeric_limits<int>::max()) {
+        // Advertise witness capabilities.
+        // The option to not set NODE_WITNESS is only used in the tests and should be removed.
         nLocalServices = ServiceFlags(nLocalServices | NODE_WITNESS);
```

Comment:
I don't think this works -- p2p_segwit launches some nodes with `-segwitheight=120` and one with segwit disabled, and expects them to follow the same blockchain. But if you set NODE_WITNESS then when the first node thinks segwit is enabled at block 120, it'll send the witness data to the other node, which will then give an error message, "AcceptBlock FAILED (unexpected-witness, ..." and fail the test.

The test case is essentially testing that upgrades work, though, which would be more useful if the non-segwit node was actually running old code, rather than just simulating it... I think the only changes needed to p2p_segwit for this are:

    * drop `test_upgrade_after_activation()`
    * in `test_getblocktemplate_before_lockin()` set `node=self.nodes[0]`, drop the for loop, and drop the dead code for `node == self.nodes[2]`.

I don't think this needs to be done in this PR though.